### PR TITLE
feat(Phind): add presence

### PR DESCRIPTION
**Pull request by GreenMan36**
## Description 
I added support for [phind.com](https://www.phind.com/), an AI powered search engine. It can take in huge pieces of information, will automatically generate a search querry based off of that and combine the search results with the input to give you an AI generated output. It will automatically detect if GPT-3 or GPT-4 should be used (This used to be manual).

Its really great for programming but it can be used for normal searches as well.

## Acknowledgements
- [x] I read the [Presence Guidelines](https://github.com/PreMiD/Presences/blob/main/.github/CONTRIBUTING.md)
- [x] I linted the code by running `yarn format`
- [x] The PR title follows the repo's [commit conventions](https://github.com/PreMiD/Presences/blob/main/.github/COMMIT_CONVENTION.md)

## Screenshots
<details>
<summary> Proof showing the creation/modification is working as expected </summary>

DiscordCanary_OEOskRl4IA
*NOTE:* The "Visit Phind" button has been removed since recording this, instead it will not show any button. This was done to follow the guidelines that state "Redirects to main page are prohibited.".
image
image

</details>

### CODE REVIEW

File: websites/P/Phind/presence.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	};
+
+	if (
+		settings.shareSearch &&
+		!settings.privateMode &&
+		document.location.pathname.startsWith("/search")
+	) {
+		presenceData.buttons = [
+			{
+				label: "🔎 Open Search Result 🔎",
```
Comment (theusaf) [inline websites/P/Phind/presence.ts:56]:
```suggestion
				label: "Open Search Result",
```

File: websites/P/Phind/presence.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+			presenceData.state = searchDetails;
+			presenceData.smallImageKey = "search";
+		}
+	}
+
+	const settings = {
+		displaySearch: await presence.getSetting("displaySearch"),
+		shareSearch: await presence.getSetting("shareSearch"),
+		privateMode: await presence.getSetting("privateMode"),
+	};
```
Comment (theusaf) [inline websites/P/Phind/presence.ts:47]: Use `await Promise.all(...)` to load multiple settings.

**Review by theusaf — CHANGES_REQUESTED** [review]

File: websites/P/Phind/presence.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
+const presence = new Presence({
+		clientId: "1106990410838065172",
+	}),
+	browsingTimestamp = Math.floor(Date.now() / 1000);
+
+presence.on("UpdateData", async () => {
+	const presenceData: PresenceData = {
+			largeImageKey: "phind-square",
```
[the remark this answers is not in the document]
Comment (GreenMan36) [reply websites/P/Phind/presence.ts:8]: Imgur seems to be down at the moment so I'll change this in a ~10 hrs.

File: websites/P/Phind/presence.ts (hunk 2 above)
Comment (GreenMan36) [reply websites/P/Phind/presence.ts]: Changed to
```typescript
const [displaySearch, shareSearch, privateMode] = await Promise.all([
	presence.getSetting("displaySearch"),
	presence.getSetting("shareSearch"),
	presence.getSetting("privateMode"),
]);
```

File: websites/P/Phind/presence.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (DarkVillager) [reply websites/P/Phind/presence.ts:8]: > 

https://imgur.com/a/U247LKm
https://i.imgur.com/fakWcYA.png
https://i.imgur.com/a5qsEbL.png

File: websites/P/Phind/presence.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+			"/default": "Viewing how to set Phind as default",
+			"/bangs": "Viewing !Bangs",
+			"/mobile": "Viewing Mobile Page",
+			"/hotkeys": "Viewing Hotkeys",
+			"/history": "Viewing Search History",
+			"/about": "Viewing About Page",
+			"/tutorial": "Viewing Tutorial",
+			"/privacy": "Viewing Privacy Policy",
+			"/terms": "Viewing Terms",
+		},
```
Comment (DarkVillager) [inline websites/P/Phind/presence.ts:26]:
```suggestion
		},
		{pathname, href} = document.location,
```

File: websites/P/Phind/presence.ts [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+		pathDetails = Object.entries(pathDetailsMap).find(([pathPrefix]) =>
+			document.location.pathname.startsWith(pathPrefix)
```
Comment (DarkVillager) [inline websites/P/Phind/presence.ts:28]:
```suggestion
			pathname.startsWith(pathPrefix)
```

File: websites/P/Phind/presence.ts [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+			presenceData.state = searchResults[
+				searchResults.length - 1
+			].querySelector("span.fw-bold.fs-3.mb-3").textContent;
+		}
+		if (shareSearch) {
+			// share search query with button
+			presenceData.buttons = [
+				{
+					label: "Open Search Result",
+					url: `https://www.phind.com/search${document.location.search}`,
```
Comment (DarkVillager) [inline websites/P/Phind/presence.ts:56]:
```suggestion
					url: href,
```

File: websites/P/Phind/presence.ts [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+			presence.getSetting("shareSearch"), // shares your search with a button
+			presence.getSetting("privateMode"), // hides the querry, button and what exact page you are on
+		]);
+
+	if (pathDetails) {
+		presenceData.details = pathDetails[1]; // display text of /path
+		presenceData.state = document.location.pathname; // display /path
+	} else if (searchResults.length > 0 && !privateMode) {
+		// we must be on /search and arn't in private mode
+		presenceData.details = "Searching";
```
Comment (DarkVillager) [inline websites/P/Phind/presence.ts:42]:
```suggestion
		presenceData.details = shareSearch ? "Searching for" : "Searching";
```

**Review by DarkVillager — CHANGES_REQUESTED** [review]

File: websites/P/Phind/presence.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (GreenMan36) [reply websites/P/Phind/presence.ts]: Alright I've 
- Moved all the consts to the top, including assets.
- Added a few comments to make it clear what the settings and some parts of the code are supposed to do.
- Made sure the private mode catches everything, including timestamps.
- Contacted the site maintainers with a request to add something so I can replace `.querySelector("span.fw-bold.fs-3.mb-3")` with something not style-specific/ more robust.

**Comment by GreenMan36** [comment]
I've been using `git push --force` to not clog up the git log with the changes I am making. Hope that's been alright.
I've made a [little note](https://github.com/PreMiD/Presences/pull/7353#discussion_r1193130455) to log what's been changed.

Feel free to check the code again.

**Comment by Extiriority** [comment]
> I've been using `git push --force` to not clog up the git log with the changes I am making. Hope that's been alright.
> 
> I've made a [little note](https://github.com/PreMiD/Presences/pull/7353#discussion_r1193130455) to log what's been changed.
> 
> 
> 
> Feel free to check the code again.

Not best practice, hard to see what you have committed

**Comment by GreenMan36** [comment]
> > I've been using `git push --force` to not clog up the git log with the changes I am making. Hope that's been alright.
> > 
> > I've made a [little note](https://github.com/PreMiD/Presences/pull/7353#discussion_r1193130455) to log what's been changed.
> > 
> > 
> > 
> > Feel free to check the code again.
> 
> Not best practice, hard to see what you have committed 

I thought so, but I haven't had much experience with contributing via pr's and havent had any experience with projects that have strict commit and pr guidelines.

How do you suggest I should commit my changes in the future while keeping the eventual merge nice and clean?

**Comment by Extiriority** [comment]
> > > I've been using `git push --force` to not clog up the git log with the changes I am making. Hope that's been alright.
> > > 
> > > I've made a [little note](https://github.com/PreMiD/Presences/pull/7353#discussion_r1193130455) to log what's been changed.
> > > 
> > > 
> > > 
> > > Feel free to check the code again.
> > 
> > Not best practice, hard to see what you have committed 
> 
> I thought so, but I haven't had much experience with contributing via pr's and havent had any experience with projects that have strict commit and pr guidelines.
> 
> How do you suggest I should commit my changes in the future while keeping the eventual merge nice and clean?

If you commit lots with little changes. Go look into 'git squash'. Else just commit & push

**Comment by theusaf** [comment]
> > > > > I've been using `git push --force` to not clog up the git log with the changes I am making. Hope that's been alright.
> > > > > I've made a [little note](https://github.com/PreMiD/Presences/pull/7353#discussion_r1193130455) to log what's been changed.
> > > > > Feel free to check the code again.
> > > > 
> > > > 
> > > > Not best practice, hard to see what you have committed
> > > 
> > > 
> > > I thought so, but I haven't had much experience with contributing via pr's and havent had any experience with projects that have strict commit and pr guidelines.
> > > How do you suggest I should commit my changes in the future while keeping the eventual merge nice and clean?
> > 
> > 
> > If you commit lots with little changes. Go look into 'git squash'. Else just commit & push
> 
> I am aware of squashing, but doesn't that still end up with multiple commits if more changes are requested? 🤔

You do not need to worry as much about squashing multiple commits or even the commit messages themselves for PRs to this repository. The eventual merge will squash all your commits into a single one. Take a look at any of the commits recently related to PRs for example.

**Comment by GreenMan36** [comment]
I have manually toggled each setting and it all seems to be working fine, even the *brand new* slideshow ^_^
I'd love another review now that thing's have changed quite a bit.

File: websites/P/Phind/presence.ts [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+	if (displayTime) presenceData.startTimestamp = browsingTimestamp;
+
+	if (pathDetails) {
+		presenceData.details = pathDetails[1]; // display text of /path
+		presenceData.state = pathname; // display /path
+		slideshow.deleteAllSlides();
+	} else if (searchResults.length > 0 && !privateMode) {
+		// we must be on /search and arn't in private mode
+		// presenceData.details = "Searching";
```
Comment (theusaf) [inline websites/P/Phind/presence.ts:49]:
```suggestion
```

File: websites/P/Phind/presence.ts [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+			pathname.startsWith(pathPrefix)
+		),
+		searchResults = document.querySelectorAll("div.row[name^='answer-']"),
+		[displayTime, displaySearch, cycleSearch, shareSearch, privateMode] =
+			await Promise.all([
+				presence.getSetting("displayTime"), // shows/hides the time spent on the site
+				presence.getSetting("displaySearch"), // shows/hides your search query text
+				presence.getSetting("cycleSearch"), // cycles through all the search query on the page
+				presence.getSetting("shareSearch"), // shares your search with a button
+				presence.getSetting("privateMode"), // hides the querry, button and what exact page you are on
```
Comment (theusaf) [inline websites/P/Phind/presence.ts:38]: The comments here may be unneeded

File: websites/P/Phind/presence.ts [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+		presenceData.state = pathname; // display /path
+		slideshow.deleteAllSlides();
+	} else if (searchResults.length > 0 && !privateMode) {
+		// we must be on /search and arn't in private mode
+		// presenceData.details = "Searching";
+		presenceData.details = displaySearch ? "Searching for:" : "Searching";
+		presenceData.smallImageKey = assets.search;
+		presenceData.smallImageText = "Searching";
+		if (shareSearch) {
+			// share search query with button
```
Comment (theusaf) [inline websites/P/Phind/presence.ts:54]: Same with these.

**Review by theusaf — CHANGES_REQUESTED** [review]

File: websites/P/Phind/presence.ts (hunk 9 above)
Comment (GreenMan36) [reply websites/P/Phind/presence.ts:38]: I think that if someone else reads these settings in the future, they may not be descriptive enough without the comments.
I'm a bit iffy about the naming anyways as I tried to make them fit into the Extension popout. In the code however, some different names might help a bit but the comments make sure they're 100% understood. Like the `cycle` and `share` consts.

But you can make the call, keep or delete?

Comment (Extiriority) [reply websites/P/Phind/presence.ts]: I think it's self explanatory enough.

>

File: websites/P/Phind/presence.ts [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+	if (displayTime) presenceData.startTimestamp = browsingTimestamp;
+
+	if (pathDetails) {
+		presenceData.details = pathDetails[1]; // display text of /path
+		presenceData.state = pathname; // display /path
+		slideshow.deleteAllSlides();
+	} else if (searchResults.length > 0 && !privateMode) {
+		// we must be on /search
+		presenceData.details = displaySearch ? "Searching for:" : "Searching";
+		presenceData.smallImageKey = assets.search;
```
Comment (theusaf) [inline websites/P/Phind/presence.ts:50]: Use standardized asset
```suggestion
		presenceData.smallImageKey = Assets.Search;
```

**Review by theusaf — CHANGES_REQUESTED** [review]

**Comment by theusaf** [comment]
Closing for inactivity. You may open a new PR after resolving the issues mentioned in this one.

**Pull request closed without merging**